### ISSUE TITLE
Proposal: Minimized UWP apps using the confirmAppClose capability should be activated when the user attempts to close them

### ISSUE THREAD

**Issue by Felix-Dev** [microsoft/microsoft-ui-xaml#2186]
# Proposal: Minimized UWP apps using the confirmAppClose capability should be activated when the user attempts to close them

## Summary
Today, when the user user attempts to close a *minimized* UWP app via the Windows Shell (Taskbar, Timeline,...) which uses the [confirmAppClose](https://docs.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations) app capability, they won't get notified by Windows in any way if additional input is required from them to properly proceed with the close request. Instead, the user will only notice that the app is not being closed, causing confusion to them as to why their close request is not actually closing the app. In fact, if further input is required from the user, they will only notice this once they themself brought the app to the foreground again. 

This behavior is described in the docs [here](https://docs.microsoft.com/en-us/uw/api/windows.ui.core.preview.systemnavigationmanagerpreview.closerequested#remarks) and the following GIF illustrates this issue using the UWP App [Quick Pad](https://www.microsoft.com/en-us/p/quick-pad-fluent-notepad-app/9pdlwqhtlsv3?rtc=2&source=lp&activetab=pivot:overviewtab) (Github repository [here](https://github.com/UWPCommunity/Quick-Pad)) which is a modern text editor alternative to the classic Notepad:

![appcloserequested-gif1](https://user-images.githubusercontent.com/1398851/77848688-cf15d800-71c6-11ea-8b0e-c6923049429f.gif)

## Rationale
Win32 apps like Notepad show a dialog upon closing the app if additional user input is required:

![win32-appcloserequested](https://user-images.githubusercontent.com/1398851/77848559-e7d1be00-71c5-11ea-8dc2-67741b51cd22.png)

While this example implementation is an improvement over the current situation for UWP apps it is still not perfect! For example, some time might have passed since the user last edited the file/data and thus they don't remember any longer if the data in question is something they need to save or not. As such, it would be quite helpful to the user if the *entire* app window is activated so they can quickly check the data referred to and can then decide if they want to save their edits or not. In other words, if the user attempts to close an UWP app like Quick Pad via the Windows Shell as shown in our example above, they would be greeted with the following screen in return if additional user input is required:

![win32-appcloserequested-2](https://user-images.githubusercontent.com/1398851/77848555-df798300-71c5-11ea-8098-eb041f25e5df.png)

**Closing the app via the Windows Shell is a deliberate action done by the user and as such should allow the app to be activated as a response if further user input is needed to proceed with the close request.**

## Scope
| Capability | Priority |
| :---------- | :------- |
| This proposal will enable end users to close a UWP app via the Windows Shell (Taskbar, Timeline,...) and be brought back into the app context automatically if further input from them is required to proceed with their close request. | Must |
| It is up to the developer to decide if the app needs to be activated and if yes, what content should be shown. | Must |

**Comment by shaheedmalik** [comment]
So, it should notify like Win32 programs do in the title bar.

**Comment by zhuxb711** [comment]
Agree

**Comment by BenJKuhn** [comment]
Felix, a little digging around shows that you're not the first person to try to use CloseReusted in this way. The documentation clearly states that the event doesn't fire when the app is minimized. While the documentation about the behavior is clear, it still leaves the question of how to properly implement this type of behavior, or whether it's possible at all. 

I've reached out to a few people internally for their thoughts, and will let you know what I hear back. I'll triage the feedback entry once I get a bit more information.

**Comment by Felix-Dev** [comment]
@shaheedmalik I thought about flashing in the taskbar which I assume you mean (see image below)
(![image](https://user-images.githubusercontent.com/1398851/78044736-5d2fc100-7375-11ea-9fed-8c2169bcd86c.png))

Two things to note here: 
1) Win32 Notepad does not flash its taskbar app button when attempting to close it while there is unsaved work.
2) As closing the app as described above is a *deliberate* user action I would prefer the app to be activated if further input is required from me. That way, the user can immediately see what further action is required from them. On the other hand, taskbar app button flashing would require another click until the user sees why their close request cannot yet proceed. 

Also worth considering: Currently, there is no taskbar app button flashing support for UWP apps, even if you use a helper Win32 process. I'm not sure if (and tbh I don't think) MS has any plans to reintroduce this method of attracting user attention considering all the other Shell notification options available (tiles, badges, notifications).

**Comment by Felix-Dev** [comment]
@BenJKuhn Would it help in tracking this issue if I move it to [Project Reunion](https://github.com/microsoft/ProjectReunion/)?

**Comment by shaheedmalik** [comment]
> @shaheedmalik I thought about flashing in the taskbar which I assume you mean (see image below)
> (![image](https://user-images.githubusercontent.com/1398851/78044736-5d2fc100-7375-11ea-9fed-8c2169bcd86c.png))
> 
> Two things to note here:
> 
> 1. Win32 Notepad does not flash its taskbar app button when attempting to close it while there is unsaved work.

You're right. I am surprised at that. It seems like use cases such as that would be a common sense reason to use it.

**Comment by BenJKuhn** [comment]
there was some good discussion about this. Generally speaking apps should be saving state that they need to during suspend, and the close event was designed for a fairly narrow use case. Not waking the app is a key component. Once suspend fires, the system assumes it can close the app without require the resources or time to wake the app again. What's being requested breaks that model. The current behavior also requires that the window be visible, though I don't know whether that's incidental or inherent to the design.

It's clear that this has broad appeal as a use case for this feature. We've taken the feedback and the owning team will consider how to address this going forward. I'll defer to @jonwis as to whether he feel this fits better under the reunion charter.

**Comment by jonwis** [comment]
Since this is more about platform functionality please do transfer it to https://github.com/microsoft/ProjectReunion/issues and we'll track it there.  I agree with @BenJKuhn , however, in that UWPs should save-suspend-terminate when prompted to do so.  Maybe there's something Project Reunion could do to help simplify save-and-restore-state behavior while we think about how to expand the windowing & lifecycle models.
